modules/jupyter-widget/src/binary-transport.js

```diff
@@ -0,0 +1,117 @@
+      return new Int16Array(data);
+    case 'uint16':
+      return new Uint16Array(data);
+    case 'float32':
+      return new Float32Array(data);
+    case 'float64':
+      return new Float64Array(data);
+    case 'int32':
+      return new Int32Array(data);
+    case 'uint32':
+      return new Uint32Array(data);
+    case 'int64':
+      return new BigInt64Array(data); // eslint-disable-line no-undef
+    case 'uint64':
+      return new BigUint64Array(data); // eslint-disable-line no-undef
+    default:
+      throw new Error(`Unrecognized dtype ${dtype}`);
+  }
+}
+
+function deserializeMatrix(arr, manager) {
+  /*
+   * Data is sent from the pydeck backend
+   * in the following format:
+   *`
+   * {'<layer ID>':
+   *     {'<accessor function name e.g. getPosition>': {
+   *        {
+   *          layer_id: <duplicate ID of layer as above>
+   *          column_name: <string name of column>
+   *          matrix: {
+   *            data: <binary data, a row major order representation of a matrix>,
+   *            shape: [<height>, <width>],
+   *            dtype: <string representation of typed array data type>
+   *          }
+   *
+   * Multiple layers and multiple accessors are possible.
+   *
+   * Objects at the JSON path `<accessor name>.matrix.data`
+   * are vectors representing a 1 x n matrix,
+   * or they can represent a m x n matrix in row major order form;
+   * shape of the matrix is given at <accessor name>.matrix.data.shape
+   */
+  if (!arr) {
+    return null;
+  }
+  const renderable = {};
+  for (const datum of arr.payload) {
+    // Each entry here represents a single column in
+    // a pandas.DataFrame arriving from the pydeck backend
+    const layerId = datum.layer_id;
+    const accessor = datum.accessor;
+    if (renderable[layerId] === undefined) {
+      renderable[layerId] = {};
+    }
+    // Creates a typed array of the numpy data as a row-major ordered matrix, with shape specified elsewhere
+    const rowMajorOrderMatrix = dtypeToTypedArray(datum.matrix.dtype, datum.matrix.data.buffer);
+    renderable[layerId][accessor] = {
+      layerId,
+      accessor,
+      columnName: datum.column_name,
+      matrix: {data: rowMajorOrderMatrix, shape: datum.matrix.shape}
+    };
+    if (renderable[layerId][accessor].matrix.data.length === 0) {
+      console.warn(`No records in accessor ${accessor} belonging to ${layerId}`); // eslint-disable-line
+    }
+  }
+  // Becomes the data stored within the widget model at `model.get('data_buffer')`
+  return renderable;
+}
+
+function constructDataProp(dataBuffer, layerId) {
+  const src = {length: 0, attributes: {}};
+  // For each accessor, we have a row major ordered matrix of data,
+  // represented as a single vector under `[accessor]matrix.data`
+  for (const accessor in dataBuffer[layerId]) {
+    const currentData = dataBuffer[layerId][accessor];
+    src.length = Math.max(currentData.matrix.shape[0], src.length);
+    src.attributes[accessor] = {
```

Comment:
Why the double conversion? Just get the objects into the right shape in `deserializeMatrix`?

Reply:
Yes that's correct, just to reshape the object.

Reply:
Why don't you get them into the correct shape the first time?

Reply:
This is reasonable. Let me modify the Python.